pipeline-model-definition/src/test/resources/failureInPostBlock.groovy

```diff
@@ -0,0 +1,49 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2018, CloudBees, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+pipeline {
+    agent none
+    stages {
+        stage("foo") {
+            steps {
+                echo "hello"
+            }
+        }
+    }
+    post {
+        always {
+            echo "SOME ${undefined} VAR"
+        }
+        success {
+            echo "MOST DEFINITELY FINISHED"
```

Comment:
Add a test where the failure occurs here.

Reply:
Not sure what that'd give us? The point of the test is just to make sure that we actually log all the exceptions thrown in the `post` conditions - I can add that test, just not sure there's a reason for it.

Reply:
It's to show that things behave as they should even in the exception throwing cases.  
For  example, we want to verify that the same behavior occurs in the `success` condition as in the `always`.  

I don't know if it matters but `failure` has a higher ordinal than `success`.  Does that mean failure will have already been considered when `success` throws an exception and as such will not execute?  Same question for `unstable` and etc.

Reply:
Gotcha. Will do. And yes, that does mean that an error in `success` will not result in `failure` being called - which has been the case all along. All we're doing here is surfacing the errors that could previously have been swallowed silently, not changing the flow behavior.

Reply:
I can see the logic of it - the `failure` condition should only fire if the stage failed not if one of the post condition steps failed - but that is not consistent.  By this same logic, the failure condition shouldn't fire if the steps all passed but there is a failure in the `always` condition (which does happen currently). 
Filed https://issues.jenkins-ci.org/browse/JENKINS-50645.